src/java.base/share/classes/java/util/WeakHashMap.java

```diff
@@ -823,7 +823,7 @@ public void remove() {
             if (modCount != expectedModCount)
                 throw new ConcurrentModificationException();
 
-            WeakHashMap.this.remove(currentKey);
+            WeakHashMap.this.remove(lastReturned.get());
```

Comment:
Unlike `currentKey`, `HashIterator` does not necessarily maintain a strong reference to the referent of `lastReturned`. If the `lastReturned` WeakReference were cleared between `lastReturned` being set, and the `remove()` call, the null key could again be erroneously removed. It would be cool to add a test case for this (maybe using large objects as keys would tempt the GC to clear the weakrefs via `jdk.test.lib.util.ForceGC`).

Since we store the `NULL_KEY` sentinel value for the null key, I believe `Entry.get()` will only return null in the case that the weakref has been cleared.  So could we instead fix this with:

```
if (lastReturned.get() != null) {
    WeakHashMap.this.remove(lastReturned.get());
}
```

?
Or, if we're worried about the ref being cleared between the null check and remove(), maybe:
```
Object lastReturnedKey = lastReturned.get();
if (lastReturnedKey != null) {
    WeakHashMap.this.remove(lastReturned.get());
}
```

Reply:
Yes, this is my concern too; lastReturned is a weak ref, and it is possible for it to have been cleared at this point, so get() can return null. I will post a more complete explanation.